{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotate LTR retrotransposons using exisiting consensus sequences\n",
    "\n",
    "#### Example data: annotate RLC_BdisC022 on chromosome 5 of *B. distachyon*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%% Define paths to input files, set parameters\n",
    "\n",
    "# path to genome\n",
    "annotendum = 'data/Bdistachyon_314_v3.0.fa'\n",
    "# path to TE consensus sequences\n",
    "consensus_seqs = 'data/RLC_BdisC022.fasta'\n",
    "\n",
    "# only retain blast hits which cover at least min_ltr_cov of the LTRs\n",
    "min_ltr_cov = 0.80\n",
    "#\n",
    "imprec = 0.2\n",
    "write_files = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%% Some classes and functions used for the annotation\n",
    "import os\n",
    "import subprocess\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "\n",
    "#%% some functions and classes\n",
    "\n",
    "class LTR_hit:\n",
    "    \n",
    "    def __init__(self, hit):\n",
    "        \n",
    "        coords = map(int, hit[7:9])\n",
    "        strt, end = min(coords), max(coords)\n",
    "        \n",
    "        self.hit = {'aln_len' : int(hit[0]),\n",
    "                    'pid' : float(hit[1]),\n",
    "                    'eval' : float(hit[10]),\n",
    "                    \n",
    "                    'TE' : hit[2].split('.')[0],\n",
    "                    'TE_strt' : int(hit[3]),\n",
    "                    'TE_end' : int(hit[4]),\n",
    "                    \n",
    "                    'contig' : hit[5],\n",
    "                    'c_strnd' : hit[6],\n",
    "                    'c_strt' : strt,\n",
    "                    'c_end' : end,\n",
    "                    'c_seq' : Seq(hit[9])\n",
    "                    }\n",
    "\n",
    "\n",
    "    def add_flanking(self, contigs, n=8):\n",
    "        \n",
    "        chrmsm = self.hit['contig']\n",
    "        strt = self.hit['c_strt']\n",
    "        end = self.hit['c_end']\n",
    "        \n",
    "        self.five_prime = contigs[chrmsm][strt-(n+1) : strt-1]\n",
    "        \n",
    "        self.three_prime = contigs[chrmsm][end : end+n]\n",
    "        \n",
    "        if self.hit['c_strnd'] == 'minus':\n",
    "            self.five_prime = self.five_prime.reverse_complement()\n",
    "            self.three_prime = self.three_prime.reverse_complement()\n",
    "            \n",
    "        \n",
    "def blastn(query, subject, outfile=False):\n",
    "    \n",
    "    cmd = ['blastn',\n",
    "           '-query', query,\n",
    "           '-subject', subject,\n",
    "           '-outfmt', '6 length pident qseqid qstart qend sseqid sstrand sstart send sseq evalue'\n",
    "           ]\n",
    "\n",
    "    proc = subprocess.Popen(cmd, stdout=subprocess.PIPE)\n",
    "    output = proc.stdout.read()\n",
    "    blast_out = output.splitlines()\n",
    "    \n",
    "    if outfile:\n",
    "        with open(outfile, 'w') as f:\n",
    "            for line in blast_out:\n",
    "                f.write(line + '\\n')\n",
    "\n",
    "    return [line.split('\\t') for line in output.splitlines()]\n",
    "\n",
    "\n",
    "def is_tsd(five_seq, three_seq, overlap=4):\n",
    "\n",
    "    tsd_max_len = len(five_seq)\n",
    "    \n",
    "    four_mers_l = []\n",
    "    four_mers_r = []\n",
    "    \n",
    "    if five_seq and three_seq:\n",
    "     \n",
    "        for i in xrange(tsd_max_len):\n",
    "            if i+4 < tsd_max_len+1:\n",
    "                four_mers_l.append(five_seq[i:i+4])\n",
    "                four_mers_r.append(three_seq[i:i+4])\n",
    "                \n",
    "    if any(x in four_mers_r for x in four_mers_l):\n",
    "        return '1'\n",
    "    else:\n",
    "        return '0'\n",
    "\n",
    "    \n",
    "def check_LTR_context(ltr_h, contigs, te_info, n=500, seq_cov=0.1):\n",
    "    \"\"\" Parameters:\n",
    "        - n: number of TE internal base pairs considered\n",
    "        - seq_cov: how much of these base pairs must be covered?\n",
    "    \"\"\"\n",
    "    \n",
    "    chrmsm = ltr_h.hit['contig']    \n",
    "    \n",
    "    coords = [ltr_h.hit['c_strt'], ltr_h.hit['c_end']]\n",
    "    strt, end = min(coords), max(coords)\n",
    "    te = ltr_h.hit['TE']\n",
    "    \n",
    "    # attenzione, strand confusioni!\n",
    "    if ltr_h.hit['c_strnd'] == 'plus':\n",
    "        left_seq = contigs[chrmsm][strt-n:strt]\n",
    "        right_seq = contigs[chrmsm][end:end+n]\n",
    "    else:\n",
    "        left_seq = contigs[chrmsm][end:end+n].reverse_complement()\n",
    "        right_seq = contigs[chrmsm][strt-n:strt].reverse_complement()\n",
    "    \n",
    "    probe_left = SeqRecord(left_seq, id='probe_left', name='probe_left', description='')\n",
    "    probe_right = SeqRecord(right_seq, id='probe_right', name='probe_right', description='')\n",
    "\n",
    "    # regions expected to be covered on consensus\n",
    "    right_exp = (te_info[te]['ltr_len'], te_info[te]['ltr_len'] + n)\n",
    "    left_exp = (te_info[te]['full_len'] - te_info[te]['ltr_len'] - n,  te_info[te]['full_len'] - te_info[te]['ltr_len'])\n",
    "\n",
    "    SeqIO.write([probe_left, probe_right], 'tmp.fasta', 'fasta')\n",
    "    blastout = blastn('tmp.fasta', te + '/' + te + '.fasta')\n",
    "    os.remove('tmp.fasta')\n",
    "    \n",
    "    best = 0\n",
    "\n",
    "    if blastout:\n",
    "        for hit in blastout:\n",
    "                  \n",
    "            strt = min(map(int, hit[7:9]))\n",
    "            end = max(map(int, hit[7:9]))\n",
    "            \n",
    "            if hit[2] == 'probe_left':\n",
    "                ol = is_overlapping(left_exp, (strt, end))               \n",
    "                if not ol:\n",
    "                    out = 'solo_LTR'\n",
    "                elif ol > best:\n",
    "                    best = ol\n",
    "                    out = 'three_prime_LTR' if ol > seq_cov else 'NA'\n",
    "                    \n",
    "    \n",
    "            elif hit[2] == 'probe_right':\n",
    "                ol = is_overlapping(right_exp, (strt, end))\n",
    "                if not ol:\n",
    "                    out = 'solo_LTR'\n",
    "                elif ol > best:\n",
    "                    best = ol\n",
    "                    out = 'five_prime_LTR' if ol > seq_cov else 'NA'\n",
    "    else:\n",
    "        out = 'solo_LTR'\n",
    "                    \n",
    "    return out\n",
    "            \n",
    "            \n",
    "def is_overlapping(a,b):\n",
    "    a_range = set(xrange(a[0], a[1]))\n",
    "    b_range = set(xrange(b[0], b[1]))\n",
    "\n",
    "    intersection = a_range & b_range\n",
    "    union = a_range | b_range\n",
    "\n",
    "    if len(intersection) == 0:\n",
    "        return False\n",
    "    else:\n",
    "        return len(intersection) / float(len(union))\n",
    "\n",
    "    \n",
    "def assign_TE_cat(a, b, te_info, imprec, min_ltr_cov=0.8):\n",
    "    \n",
    "    \"\"\" Determine TE insertion category:\n",
    "    \n",
    "    - full length\n",
    "    - solo LTR\n",
    "    - single LTR\n",
    "    - tsd yes or no\n",
    "    \n",
    "    1 in the output means TSD is present\n",
    "    \"\"\"\n",
    "    \n",
    "    te = a.hit['TE']\n",
    "    if a.hit['aln_len'] < len(ltrs[te]) * min_ltr_cov:\n",
    "        return 'trunc', 0, 0\n",
    "    \n",
    "    aa_tsd = is_tsd(a.five_prime, a.three_prime)      \n",
    "    a_cntxt = check_LTR_context(a, contigs, te_info)\n",
    "    \n",
    "    if not b or te != b.hit['TE']: \n",
    "        return 'single', aa_tsd, a_cntxt\n",
    "      \n",
    "    else:\n",
    "        b_cntxt = check_LTR_context(b, contigs, te_info)\n",
    "        internal = te_info[te]['full_len'] -(2*te_info[te]['ltr_len'])\n",
    "        interval = b.hit['c_strt'] - a.hit['c_end']\n",
    "        ab_tsd = is_tsd(a.five_prime, b.three_prime)\n",
    "        cntxt = a_cntxt + '-' + b_cntxt\n",
    "        \n",
    "        if (1-imprec)*internal < interval < (1+imprec)*internal and a.hit['c_strnd'] == b.hit['c_strnd']:\n",
    "            return 'paired', ab_tsd, cntxt\n",
    "        \n",
    "        else:\n",
    "            return  'single', aa_tsd, a_cntxt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%% Load sequence data and get the LTR sequence for each element\n",
    "contigs = {seq_record.id: seq_record.seq for seq_record in SeqIO.parse(annotendum, \"fasta\")}\n",
    "\n",
    "tes = {seq_record.id: seq_record for seq_record in SeqIO.parse(consensus_seqs, \"fasta\")}\n",
    "basename = consensus_seqs.split('/')[-1]\n",
    "ltrs = {}\n",
    "\n",
    "os.mkdir('tmp')\n",
    "os.chdir('tmp')\n",
    "\n",
    "for seq in tes:\n",
    "    \n",
    "    SeqIO.write(tes[seq], seq + '.fasta', 'fasta')\n",
    "    \n",
    "    selfblast = blastn(seq + '.fasta', seq + '.fasta')[1:]\n",
    "    selfblast.sort(key=lambda x: (int(x[7])))  \n",
    "    \n",
    "    ltrs[seq] = SeqRecord(Seq(selfblast[0][-2]), id=seq, name=seq, description='')      \n",
    "    os.remove(seq + '.fasta')\n",
    "    \n",
    "os.chdir('..')\n",
    "os.rmdir('tmp')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Annotating RLC_BdisC022\n"
     ]
    }
   ],
   "source": [
    "#%% Blast LTRs and parse hits\n",
    "\n",
    "blast_d = {}\n",
    "te_info = {}\n",
    "overlaps = []\n",
    "\n",
    "\n",
    "for te in tes:\n",
    "    \n",
    "    \"\"\" Blast LTR of each TE family against the assembly, \n",
    "    create LTR_hit object and extract sequences flanking the hit\n",
    "    \"\"\"\n",
    "    \n",
    "    print 'Annotating ' + te\n",
    "\n",
    "    if te in ignore:\n",
    "        continue\n",
    "    \n",
    "    os.mkdir(te)\n",
    "\n",
    "    SeqIO.write(tes[te], te + '/' + te + '.fasta', 'fasta')\n",
    "    SeqIO.write(ltrs[te], te + '/' + te + '.LTR.fasta', 'fasta')\n",
    "    full_len = len(tes[te])\n",
    "    ltr_len = len(ltrs[te])    \n",
    "    te_info[te] = {'full_len' : full_len, 'ltr_len' : ltr_len}\n",
    "\n",
    "\n",
    "    \"\"\" Blast LTR against targets, create dictionary with blast output with chromosomes as keys\n",
    "    \"\"\"\n",
    "    stats[te] = {'retained' : 0 , 'trunc' : 0, 'overlap' : 0}\n",
    "    blast_out = blastn(te + '/' + te + '.LTR.fasta', annotendum) \n",
    "    \n",
    "    for hit in blast_out:\n",
    "        \n",
    "        # Only consider hits covering at least 80% of the LTR\n",
    "        aln_len = int(hit[0])\n",
    "#        if aln_len < min_ltr_cov*ltr_len:\n",
    "#            stats[te]['trunc'] +=1\n",
    "#            continue\n",
    "                    \n",
    "        chrmsm = hit[5]\n",
    "        if chrmsm not in blast_d:\n",
    "            blast_d[chrmsm] = {}\n",
    "    \n",
    "        coords = map(int, hit[7:9])\n",
    "        strt, end = min(coords), max(coords)\n",
    "        \n",
    "        ltr_h = LTR_hit(hit)\n",
    "        ltr_h.add_flanking(contigs)\n",
    "        \n",
    "        \n",
    "        # overlapping coordinates: keep hit with higher PID\n",
    "        overlap = False\n",
    "        \n",
    "        worse_hits_to_remove = []\n",
    "        for region in blast_d[chrmsm]:\n",
    "            ol = is_overlapping((strt,end), region)\n",
    "            \n",
    "            if ol: \n",
    "                overlap = True\n",
    "                if ltr_h.hit['eval'] < blast_d[chrmsm][region].hit['eval']:\n",
    "                    worse_hits_to_remove.append(region)\n",
    "        \n",
    "        if worse_hits_to_remove:\n",
    "            for h in worse_hits_to_remove:\n",
    "                blast_d[chrmsm].pop(h)\n",
    "            blast_d[chrmsm][(strt, end)] = ltr_h\n",
    "                \n",
    "        elif overlap == False:\n",
    "            blast_d[chrmsm][(strt, end)] = ltr_h\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%% Identify full length and solo copies, write gff and fasta files\n",
    "annot = []\n",
    "seqs = {}\n",
    "ltr_seqs = {}\n",
    "annot_stats = {}\n",
    "\n",
    "for chrmsm in blast_d:\n",
    "        \n",
    "        \"\"\" Order blast hits and compare adjacent hits in order\n",
    "        to distinguish solo from full length elements,\n",
    "        and to identify overlapping annotations\n",
    "        \"\"\"\n",
    "        \n",
    "        coords = blast_d[chrmsm].keys()\n",
    "        coords.sort(key=lambda x: (int(x[0])))\n",
    "\n",
    "        a = False\n",
    "        \n",
    "        for i,x  in enumerate(coords):\n",
    "            \n",
    "            if not a: # initiate contig\n",
    "                a = blast_d[chrmsm][x]\n",
    "                \n",
    "                if len(coords) == 1: # single hit in the contig\n",
    "                    b = False\n",
    "                    cat, tsd, ltr_cntxt = assign_TE_cat(a, b, te_info, imprec)\n",
    "                    \n",
    "                else:\n",
    "                    continue\n",
    "            \n",
    "            b = blast_d[chrmsm][x]\n",
    "            cat, tsd, ltr_cntxt = assign_TE_cat(a, b, te_info, imprec)\n",
    "            \n",
    "        \n",
    "            \"\"\" Put together output info\n",
    "            \"\"\"\n",
    "            TE = a.hit['TE']\n",
    "            strt = a.hit['c_strt']\n",
    "            flanking_left = a.five_prime\n",
    "            # careful with fasta header names, avoid - and :\n",
    "            name = TE + '_' + chrmsm + '_' + str(strt)\n",
    "            \n",
    "            if TE not in annot_stats:\n",
    "                annot_stats[TE] = {'paired' : 0, 'single' : 0, 'solo_LTR' : 0, 'full' : 0, 'other' : 0, 'trunc':0}\n",
    "                seqs[TE] = []\n",
    "                ltr_seqs[TE] = []\n",
    "            \n",
    "            # strand\n",
    "            strand = '+' if a.hit['c_strnd'] == 'plus' else '-'\n",
    "            \n",
    "            # end coordinate\n",
    "            if cat == 'paired':\n",
    "                flanking_right = b.three_prime\n",
    "                end = b.hit['c_end']\n",
    "            else:\n",
    "                flanking_right = a.three_prime\n",
    "                end = a.hit['c_end']\n",
    "                \n",
    "\n",
    "            # Categorize and get LTR sequences \n",
    "            if cat == 'single':\n",
    "                if ltr_cntxt == 'solo_LTR'and tsd == '1':\n",
    "                    subcat = 'solo_LTR'\n",
    "                else:\n",
    "                    subcat = 'other'\n",
    "                    \n",
    "                ltr_rec = SeqRecord(a.hit['c_seq'], id=name, name=name, description=cat + '_' + subcat)\n",
    "                ltr_seqs[TE].append(ltr_rec)\n",
    "                \n",
    "                \n",
    "            if cat == 'paired':\n",
    "                \n",
    "                if strand == '-' and ltr_cntxt in ['three_prime_LTR-five_prime_LTR', 'three_prime_LTR-NA', 'NA-five_prime_LTR'] or \\\n",
    "                   strand == '+' and ltr_cntxt in ['five_prime_LTR-three_prime_LTR', 'NA-three_prime_LTR', 'five_prime_LTR-NA']:\n",
    "                       \n",
    "                    subcat = 'full' if tsd == '1' else 'other'\n",
    "                    \n",
    "                else:\n",
    "                    subcat = 'other'\n",
    "    \n",
    "                ltr_recA = SeqRecord(a.hit['c_seq'], id=name + '_A', name=name + '_A', description=cat + '_' + subcat)\n",
    "                ltr_recB = SeqRecord(b.hit['c_seq'], id=name + '_B', name=name + '_B', description=cat + '_' + subcat)\n",
    "                ltr_seqs[TE] += [ltr_recA, ltr_recB]\n",
    "                \n",
    "            if cat == 'trunc':\n",
    "                subcat = 'trunc'\n",
    "                \n",
    "                \n",
    "            # TE sequence and length\n",
    "            seq = contigs[chrmsm][strt-1 : end]\n",
    "            aln_len = len(seq)\n",
    "            if strand == '-':\n",
    "                seq = seq.reverse_complement()\n",
    "            rec = SeqRecord(seq, id=name, name=name, description=cat+ '_' + subcat)\n",
    "            seqs[TE].append(rec)\n",
    "                \n",
    "\n",
    "            \"\"\" Write reference hits to gff\n",
    "            \"\"\"\n",
    "            info = 'Name=%s,fam=%s,cat=%s,subcat=%s,tsd=%s,ltr_context=%s,five_flanking=%s,three_flanking=%s,aln_len=%s' % \\\n",
    "                    (name, TE, cat, subcat, tsd, ltr_cntxt, flanking_left, flanking_right, aln_len)\n",
    "                    \n",
    "            annot.append([chrmsm, 'detettore', 'LTR-RT', str(strt), str(end), '.', strand, '.', info])\n",
    "\n",
    "            annot_stats[TE][cat] += 1 \n",
    "            annot_stats[TE][subcat] += 1\n",
    "            \n",
    "            # Skip next entry if it was part of a full element\n",
    "            if cat  == 'paired':\n",
    "                a = False\n",
    "            else:\n",
    "                a = b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write files\n",
    "                \n",
    "stats_header = ['TE', 'paired', 'single', 'full', 'solo_LTR', 'other']\n",
    "              \n",
    "if write_files:\n",
    "    annot.sort(key=lambda x: (x[0], int(x[3])))\n",
    "    with open('LTR-RTs.dez2018.gff', 'w') as out:\n",
    "        for a in annot:\n",
    "            out.write('\\t'.join(a)+'\\n')\n",
    "    \n",
    "    for te in seqs:\n",
    "        SeqIO.write(seqs[te], te + '/' + te + '.all_seqs.fasta', 'fasta')\n",
    "        SeqIO.write(ltr_seqs[te], te + '/' + te + '.ltr_seqs.fasta', 'fasta')\n",
    "        \n",
    "    \n",
    "    with open('annotation_counts.tsv', 'w') as f:\n",
    "        f.write('\\t'.join(stats_header) + '\\n')\n",
    "        for te in annot_stats:\n",
    "            outline = [te] + [str(annot_stats[te][c]) for c in stats_header[1:]]\n",
    "            f.write('\\t'.join(outline) + '\\n')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
